app/Libraries/ForumSearch.php

```diff
@@ -0,0 +1,190 @@
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace App\Libraries;
+
+use App\Libraries\Elasticsearch\BoolQuery;
+use App\Libraries\Elasticsearch\HasChildQuery;
+use App\Libraries\Elasticsearch\Highlight;
+use App\Libraries\Elasticsearch\Search;
+use App\Libraries\Elasticsearch\SearchResponse;
+use App\Models\Forum\Forum;
+use App\Models\Forum\Post;
+use App\Models\Forum\Topic;
+use App\Models\User;
+
+// FIXME: remove ArrayAccess after refactored
+class ForumSearch extends Search implements \ArrayAccess
+{
+    const HIGHLIGHT_FRAGMENT_SIZE = 50;
+
+    protected $includeSubforums;
+    protected $queryString;
+    protected $username;
+    protected $forumId;
+
+    public function __construct(array $options = [])
+    {
+        parent::__construct(Post::esIndexName(), $options);
+
+        $this->queryString = $options['query'];
+        $this->includeSubforums = get_bool($options['includeSubforums'] ?? false);
+        $this->username = presence($options['username'] ?? null);
+        $this->forumId = get_int($options['forumId'] ?? null);
+        $this->topicId = get_int($options['topicId'] ?? null);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toArray() : array
+    {
+        $match = ['query_string' => [
+            'fields' => ['search_content'],
+            'query' => $this->queryString,
+        ]];
+
+        $query = (new BoolQuery())
+            ->must(static::firstPostQuery()->toArray())
+            ->should($this->childQuery()->toArray())
+            ->shouldMatch(1)
+            ->filter(['term' => ['type' => 'topics']]);
+
+        // skip the topic search if doing a username; needs a more complicated
+        // query to accurately filter the results which isn't implemented yet.
+        if (!isset($this->username)) {
+            $query->should($match);
+        }
+
+        if (isset($this->forumId)) {
+            $forumIds = $this->includeSubforums
+                ? Forum::findOrFail($this->forumId)->allSubForums()
+                : [$this->forumId];
+
+            $query->filter(['terms' => ['forum_id' => $forumIds]]);
+        }
+
+        if (isset($this->topicId)) {
+            $query->filter(['term' => ['topic_id' => $this->topicId]]);
+        }
+
+        $this->query($query);
+
+        return parent::toArray();
+    }
+
+    private function childQuery() : HasChildQuery
+    {
+        $query = (new BoolQuery())
+            ->must(['query_string' => [
+                'fields' => ['search_content'],
+                'query' => $this->queryString,
+            ]]);
+
+        if (isset($this->username)) {
+            $user = User::lookup($this->username);
+            $query->filter(['term' => ['poster_id' => $user ? $user->user_id : -1]]);
+        }
+
+        return (new HasChildQuery('posts', 'posts'))
+            ->size(3)
```

Comment:
may want to increase limit especially for topic search...? Or maybe something completely different for topic search with pagination and stuff.

Reply:
Or let's do this later :best: